### ISSUE TITLE
resources#relativePath does not honour path casing strategy

### ISSUE THREAD

**Issue by bpasero** [microsoft/vscode#108256]
It seems to me that [`resources.relativePath`](https://github.com/microsoft/vscode/blob/6e022766d135689a6a5f92a4f3d9ebf97b238dd0/src/vs/base/common/resources.ts#L228-L228) ignores the fact that path casing should be ignored depending on the underlying file system. As you can see here:

https://github.com/microsoft/vscode/blob/6e022766d135689a6a5f92a4f3d9ebf97b238dd0/src/vs/base/common/resources.ts#L232-L235

Extracted from https://github.com/microsoft/vscode/issues/107774

> I am actually no longer so sure that path casing is ignored, @jrieken is that handling for file URI intentional?

Only @aeschli knows, he wrote those utils and I didn't change their semantics when moving them to `IExtUri`. But, yeah it seems that there is some scheme and platform specific handling in there and I don't think that is good.

_Originally posted by @jrieken in https://github.com/microsoft/vscode/issues/107774#issuecomment-705017638_

**Comment by jrieken** [comment]
Not many references, should be a manageable risk

```
src/vs/base/common/resources.ts
  103, 2: 	relativePath(from: URI, to: URI): string | undefined;
  228, 2: 	relativePath(from: URI, to: URI): string | undefined {
  366, 36: relativePath = extUri.relativePath.bind(extUri);

src/vs/base/test/common/resources.test.ts
  247, 21: 		assert.equal(util.relativePath(u1, u2), expectedPath, `from ${u1.toString()} to ${u2.toString()}`);

src/vs/platform/files/browser/indexedDBFileSystemProvider.ts
  148, 25: 	const path = extUri.relativePath(resource, keyResource);

src/vs/platform/userDataSync/common/snippetsSync.ts
  493, 29: const key = this.extUri.relativePath(this.snippetsFolder, resource)!;

src/vs/workbench/contrib/scm/browser/activity.ts
  78, 48: uriIdentityService.extUri.relativePath(root, resource);

src/vs/workbench/services/userData/common/fileUserDataProvider.ts
  145, 37: relativePath = this.extUri.relativePath(this.backupHome, userDataResource);
  149, 36: relativePath = this.extUri.relativePath(this.userDataHome, userDataResource)!;
  155, 37: relativePath = this.extUri.relativePath(this.fileSystemUserDataHome, fileSystemResource);
  159, 37: relativePath = this.extUri.relativePath(this.fileSystemBackupsHome, fileSystemResource);

```

**Comment by aeschli** [comment]
All the utils in `resources` have special handling of `file` uris. All is converted to local paths (fsPath) and the local variant of the path library is used. 
That helped with the path casing but also with the root handing (e.g. `c:\` is preserved).

That only works when the `file` URI is really from a file of the local filesystem. And that's normally the case. Remote files should always show up as `vscode-remote://` on the local process.

I'm not opposed to remove the special `file` handing and give up on the root handing.

But maybe for #107774, maybe the issue is that resolving should happen on the remote side?

**Comment by bpasero** [comment]
I would expect this to work:

```ts
const uriA = URI.file('/foo/BAR/somefile.txt');
const uriB = URI.file('/foo/bar');
console.log(this.uriIdentityService.extUri.relativePath(uriB, uriA)) // prints "someFile.txt" on macOS/Windows and nothing on Linux
```

Given the check for `file` scheme here:

https://github.com/microsoft/vscode/blob/6e022766d135689a6a5f92a4f3d9ebf97b238dd0/src/vs/base/common/resources.ts#L232-L235

I think the later ignorance of path casing is not applied:

https://github.com/microsoft/vscode/blob/6e022766d135689a6a5f92a4f3d9ebf97b238dd0/src/vs/base/common/resources.ts#L237-L237

**Comment by bpasero** [comment]
To follow up: node.js itself seems to be ignoring path-case differences in their `path.relative` method, which I think is bad. I would never think we can rely on their implementation with regards to ignoring casing or not because we do know exactly from the registered filesystem capabilities.

**Comment by aeschli** [comment]
I agree. If we know the case sensitivity of a scheme (as it's the case with extURI), we should not use the `file` special case.

@jrieken Assigning to you, extURI is your baby.

**Comment by jrieken** [comment]
> @jrieken Assigning to you, extURI is your baby.

Nope - it's your code, your tests that I have moved into this interface

**Comment by aeschli** [comment]
`resource.relative` now uses the `unbiased` extUri which means ignorePathCasing is hardcoded to false. That will be the case also for `file` on windows.
Not sure if that was intentional. But if I were to remove the `file` handling and use `_ignorePathCasing`  we will see some change in behavior.

**Comment by jrieken** [comment]
> Not sure if that was intentional.

Yes - to recap: all implicit case ignore logic is gone. The default is to be case-sensitive (which is correct because we use canonical uris) Also, `IUriIdentityService#extUri` is the uri-utility that knows when to ignore/honour path casing.  


> But if I were to remove the file handling and use _ignorePathCasing we will see some change in behavior.

Yeah, that's usually the desire when filing a bug.

**Comment by aeschli** [comment]
I prepared a change to no longer have a `file` special case, and to only follow the `_ignorePathCasing` (https://github.com/microsoft/vscode/pull/108796).

This unconverted some issues where `relativePath` is used on two Windows paths with different drive letter casing. That results in relative paths like `../../../c:/foo/bar/hello`.

I went through all the references of `resources.relativePath` and I'm afraid almost none of them are guaranteed canonical and might be prone to the same issue.

An example:
https://github.com/microsoft/vscode/blob/627fa8d42529364741784dbb71db8a07b71e5bc9/src/vs/base/common/labels.ts#L42
where the resource might have been created just a few lines above.

**Comment by bpasero** [comment]
@aeschli I would have expected that the drive-letter casing is not relevant given we ignore casing on Windows?

**Comment by aeschli** [comment]
@bpasero See https://github.com/microsoft/vscode/issues/108256#issuecomment-709436838: since a while `resources.relativePath` maps to `extUri.relativePath` which is case sensitive for all schemas/platforms. 
We never ran into any issues as for `file` schems we didn't honor the path casing strategy (the topic of this issue), 
Now with my proposed fix (#108796), this is uncovered.

We must go through all usages of `resources.relativePath` and either
a. change it to `extUriBiasedIgnorePathCase.relativePath` to get the old behavior
b. Make sure that the URIs that are passed in are canonical uris

Also a possibility: 
c. we change `resources.relativePath` to `extUriBiasedIgnorePathCase.relativePath`.
d. we ignore case for drive letters only (we already have code in relativePath to detect windows drive segments)

(b.) is the preferred solution, but requires some work, e.g.  #108793

**Comment by bpasero** [comment]
Yes, we need to ensure the canonical URI is used unless people are using `IURIIdentityService` which is really what I was suggesting and actually prefer. Why would people bother with canonical URIs if you can just use `IURIIDentityService`. I really think NO ONE should be producing canonical URIs, only in very specific places as we do today where we create text models or editors.

Anyone else should use `IURIIdentityService` when dealing with URIs and then this should not be an issue at all.
